code/iaas/lb-instance/src/main/java/io/cattle/platform/lb/instance/process/LoadBalancerUpdateConfig.java

```diff
@@ -177,45 +190,69 @@ public boolean evaluate(Instance obj) {
             List<? extends Instance> lbInstances = lbInstanceManager.getLoadBalancerInstances(lb);
-            for (Instance lbInstance : lbInstances) {
-                Map<Integer, Port> portsToCreate = new HashMap<Integer, Port>();
-                Map<Integer, Port> portsToRemove = new HashMap<Integer, Port>();
-                for (Port port : objectManager.children(lbInstance, Port.class)) {
-                    if (listenerPorts.contains(port.getPublicPort())) {
-                        portsToCreate.put(port.getPublicPort(), port);
-                    } else {
-                        portsToRemove.put(port.getPublicPort(), port);
-                    }
-                }
-                for (Integer listenerPort : listenerPorts) {
-                    if (portsToCreate.containsKey(listenerPort)) {
-                        continue;
-                    }
-                    Nic primaryNic = networkDao.getPrimaryNic(lbInstance.getId());
-                    if (primaryNic == null) {
-                        continue;
-                    }
-                    IpAddress ipAddress = ipAddressDao.getPrimaryIpAddress(primaryNic);
-                    if (ipAddress == null) {
-                        continue;
+            // surround by lock
+            for (final Instance lbInstance : lbInstances) {
+                final boolean hasActiveInstances = (targetDao.getLoadBalancerActiveInstanceTargets(lb.getId()).size() + targetDao
+                        .getLoadBalancerActiveIpTargets(lb.getId()).size()) > 0;
+                final Map<Integer, Port> portsToCreate = new HashMap<Integer, Port>();
+                final Map<Integer, Port> portsToRemove = new HashMap<Integer, Port>();
+                lockManager.lock(new LoadBalancerInstancePortLock(lbInstance), new LockCallbackNoReturn() {
+                    @Override
+                    public void doWithLockNoResult() {
+                        Map<Integer, Port> existingPorts = new HashMap<Integer, Port>();
+                        for (Port port : objectManager.children(lbInstance, Port.class)) {
+                            if (port.getRemoved() == null && !port.getState().equals(CommonStatesConstants.REMOVED)) {
+                                existingPorts.put(port.getPublicPort(), port);
+                            }
+                        }
+                        for (Port port : existingPorts.values()) {
+                            if (listenerPorts.contains(port.getPublicPort()) && hasActiveInstances) {
+                                portsToCreate.put(port.getPublicPort(), port);
+                            } else {
+                                portsToRemove.put(port.getPublicPort(), port);
+                            }
+                        }
+
+                        if (hasActiveInstances) {
+                            for (Integer listenerPort : listenerPorts) {
+                                if (existingPorts.containsKey(listenerPort)) {
```

Comment:
Just out of curiosity.  Previously we were checking portsToCreate.  In the new changes, we're checking existingPorts which is a superset of portsToCreate.